### ISSUE TITLE
Discussion: Redux bindings for AngularJS (ng-redux)

### ISSUE THREAD

**Issue by wbuchwalter** [rackt/redux#387]
As requested by @gaearon a few days ago, i'm opening an issue to discuss Redux integration with Angular, specifically [ng-redux](https://github.com/wbuchwalter/ng-redux/).

- The current version is working with Redux@1.0.0
- Basic documentation is done
- Tests are presents
- Works well with [reselect](https://github.com/faassen/reselect)

You can see a working example project here: [tsRedux](https://github.com/wbuchwalter/tsRedux/).

Any feedback is welcome.

**Comment by jbalboni** [comment]
I think it looks great. I would really like to see an example of using it with routing (ui-router, mainly). I'm just starting to learn about Redux and that would help me out.

**Comment by wbuchwalter** [comment]
Hey @jbalboni , I didn't had the occasion to play with ui-router yet, I will try to find time for that (no guarantee).

I've also started [ng2-redux](https://github.com/wbuchwalter/ng2-redux): bindings for Angular 2, with an example counter app using it.

**Comment by elgerlambert** [comment]
Hi @wbuchwalter,

Good to see the progress you've made since our initial discussions on Slack. I was wondering why you've opted to stick to using a callback rather than passing in the scope/this, like so:

```js
class TodoLoaderController {

  constructor($ngRedux) {
    this.todos = [];
    // $ngRedux.connect will set this.todos whenever it changes and take care of
    // unsubscribing when the controller is $destroyed.
    $ngRedux.connect(state => state.todos, this);
  }

  [...]
}
```
Is there a use-case whereby you would like to do other things within the connect callback other then assigning the state to the scope/this?

The current implementation burdens the developer to remember to teardown and unsubscribe from connect. By passing in the scope/this you can take care of unsubscribing within $ngRedux by listening to `this.$on("$destroy", ..` (or by wrapping the `this.destroy` function I guess - I'm unfamiliar with the use of es6 classes in angular, is that common now?)

**Comment by wbuchwalter** [comment]
Hey @elgerlambert!

Yea, I was split between this two possibilities, and ended up with the callback for more flexibility.
For example:
* If there is a timeout in one of your middlewares/action creators (or if you use something else than $http/$q), you will need to trigger a $apply somewhere. The callback gives you a hook to do that.
* If you have some business logic specific to a controller, you could do that here instead of creating a custom selector/reducer
* When dealing with a lot of reducers, you may very well end up with recurring properties name (e.g: todos.map and counter.map), so both would be merged on the same property of your scope.
* You could pass something else than $scope, so I need to detect that possibility to know when to listen to $destroy:
```JS
  constructor($ngRedux) {
      this.todos = {};
      $ngRedux.connect(state => state.todos.map, this.todos);
    }
```


But those are mostly edge cases / consequence of bad designed code...
The truth is I only know how I (and probably you) would structure a redux app with angular, but I'm not totally confident this would also apply to a majority of other people, so I opted for the more versatile option.

With that said, I could also allow both a callback or an object to be passed to ```connect``` and modify the behavior based of ngRedux on that, that would allow for both usages...

Let me know what you think!

**Comment by elgerlambert** [comment]
1. I feel the responsibility to ensure a $digest cycle is triggered lies with $ngRedux; I would expect it to take care of this. In that respect it's similar to using setTimeout or $timeout; with the former I know I have to handle the $digest cycle, but when you're using a "$" service you expect this to be taken care of. You should check whether a $digest cycle is already in progress (triggered by e.g. $http/$q) and if not trigger one. My suggestion would be to wrap the store's subscribe function with a function that takes care of this.
2. This is the point where I'm not sure if there is a valid case, there may well be - I'm not sure. The way you've phrased this second point though unfortunately raises my concerns that the extra flexibility  offered by a callback only makes it easier to do things that would be considered bad practise, such as triggering side-effects.
3. This is easily solved; the return of a selector should be an object rather then the "state slice". This allows for greater flexibility (since it allows the consumer/controller to define it's own keys) and removes the need to pass in an Array of selectors if you want multiple values from the store. It'll also get you closer to react-redux's API. See snippet below.
4. You *could* allow this, but I don't see why you *should*. I would create an Invariant to check if the passed in object is in fact a $scope object and throw if not.

```js
$ngRedux.connect(state => ({
  todos: state.todos.map,
  counter: state.counter.map
}), this);
```

As a side note, wouldn't it be useful to expose the store's dispatch directly e.g. `$ngRedux.dispatch()`? How do you do this now, `$ngRedux.getStore.dispatch()`?

**Comment by wbuchwalter** [comment]
The first major release of `ng-redux` was just released (`2.0.0` to be consistent with Redux), correcting most of the things discussed here.
We also added injectable middlewares, support for time machine and better doc.
Thanks a lot to @elgerlambert for all the help.

**Comment by gaearon** [comment]
Closing, as there have been feedback and progress, and there is nothing actionable for Redux team here. Thank you for working on this!
